text/0112-server-runtime.md

```diff
@@ -0,0 +1,331 @@
+```
+
+### `@lwc/engine-core`
+
+This packages exposes the following platform-agnostic APIs:
+
+- `LightningElement`
+- `api`
+- `track`
+- `readonly`
+- `wire`
+- `setFeatureFlag`
+- `getComponentDef`
+- `isComponentConstructor`
+- `getComponentConstructor`
+- `unwrap`
+- `registerTemplate`
+- `registerComponent`
+- `registerDecorators`
+- `sanitizeAttribute`
+
+The DOM APIs used by the rendered engine are injected by the runtime depending on the environment. A list of all the DOM APIs the engine depends upon can be found in the [DOM APIs usage](#dom-apis-usage) section in the Appendix.
+
+### `@lwc/engine-dom`
+
+This package exposes the following APIs:
+
+- `createElement` + reaction hooks
+- `buildCustomElementConstructor`
+- `isNodeFromTemplate`
+
+This package injects the native DOM APIs into the `@lwc/engine-core` rendering engine.
+
+### `@lwc/engine-server`
+
+This package exposes the following APIs:
+
+- `createElement(name: string, options: { is: typeof LightningElement }): ServerHTMLElement`: This method creates a new LWC component tree. It follows the same signature as the `createElement` API from `@lwc/engine-dom`. Instead of returning a native `HTMLElement`, this method returns a `ServerHTMLElement` with the public properties, aria reflected properties and HTML global attributed.
+- `renderToString(element: ServerHTMLElement): string`: This method creates an LWC component tree synchronously and serialize it to string. It accepts a single parameter, a `ServerHTMLElement` returned by `createElement` and returns the serialized string.
+
+This package injects mock DOM APIs in the `@lwc/engine-core` rendering engine. Those DOM APIs produces a lightweight DOM structure that can be serialized into a string by the `renderToString` method. As described in the Appendix, this package is also in charge of attaching on the global object a mock `CustomEvent`.
+
+## Drawbacks
+
+- Requires a complete rewrite of the current `@lwc/engine`
+- Might require substantial changes to the existing tools (jest preset, rollup plugins, ...) to load the right engine depending on the environment.
+
+## Alternatives
+
+### Using a DOM implementation in JavaScript before evaluating the engine
+
+Prior the evaluation of the LWC engine, we would evaluate a DOM implementation write in JavaScript (like [jsdom](https://github.com/jsdom/jsdom) or [domino](https://github.com/fgnass/domino)). The compelling aspect of this approach is that it requires almost no change to the engine to work.
+
+There are multiple drawbacks with this approach:
+
+- Performance: The engine only relies on a limited set of well-known APIs, leveraging a full DOM implementation for SSR would greatly reduce the SSR performance.
+- Predictability: By attaching the DOM interfaces on the global object, those APIs are not only exposed to the engine but also to the component author. Exposing such APIs to the component author might bring unexpected behavior when component code runs on the server.
```

Comment:
I would also add scalability. The mocked APIs would have to be kept in sync with what's actually being used in the engine. Also the engine still wouldn't be reusable in other environments without a corresponding "mock" layer.

Reply:
I don't think it is the case. There is only a well-defined set of DOM APIs the engine requires to operate. Each runtime (dom and server) is in charge of implementing this contract by passing an implementation of those APIs to the engine.

Reply:
Maybe it has stabilized by now but when I took that approach I had to add/change multiple mocked APIs during the couple months I touched that code.

Reply:
I don't think that will be that challenging, we will own both packages, and we can keep them in sync.